File: frontend/app/types/achievement.ts
export interface Achievement {
  id: string
  title: string
  issuer?: string | null
  description?: string | null
  category?: string | null
  achievedAt?: string | null
  credentialId?: string | null
  externalUrl?: string | null
  certificateFile?: string | null
  displayOrder: number
  isVisible: boolean
}

export interface RawAchievement {
  id: string
  title: string
  issuer?: string | null
  description?: string | null
  category?: string | null
  achieved_at?: string | null
  achievedAt?: string | null
  credential_id?: string | null
  credentialId?: string | null
  external_url?: string | null
  externalUrl?: string | null
  certificate_file?: string | null
  certificateFile?: string | null
  display_order?: number
  displayOrder?: number
  is_visible?: boolean
  isVisible?: boolean
}

export function normalizeAchievement(raw: RawAchievement): Achievement {
  return {
    id: raw.id,
    title: raw.title,
    issuer: raw.issuer ?? null,
    description: raw.description ?? null,
    category: raw.category ?? null,
    achievedAt: raw.achievedAt ?? raw.achieved_at ?? null,
    credentialId: raw.credentialId ?? raw.credential_id ?? null,
    externalUrl: raw.externalUrl ?? raw.external_url ?? null,
    certificateFile: raw.certificateFile ?? raw.certificate_file ?? null,
    displayOrder: raw.displayOrder ?? raw.display_order ?? 0,
    isVisible: raw.isVisible ?? raw.is_visible ?? true,
  }
}

File: backend/internal/models/achievement.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Achievement struct {
	ID                uuid.UUID  `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	UserID            uuid.UUID  `gorm:"type:uuid;not null" json:"userId"`
	Title             string     `gorm:"type:varchar(255);not null" json:"title"`
	Issuer            *string    `gorm:"type:varchar(255)" json:"issuer"`
	Description       *string    `gorm:"type:text" json:"description"`
	Category          string     `gorm:"type:varchar(30);not null" json:"category"`
	Level             *string    `gorm:"type:varchar(50)" json:"level"`
	AchievedAt        *time.Time `gorm:"type:date" json:"achievedAt"`
	CredentialID      *string    `gorm:"type:varchar(255)" json:"credentialId"`
	ExternalURL       *string    `gorm:"type:text" json:"externalUrl"`
	CertificateFileID *uuid.UUID `gorm:"type:uuid" json:"certificateFileId"`
	DisplayOrder      int        `gorm:"default:0" json:"displayOrder"`
	IsVisible         bool       `gorm:"default:true" json:"isVisible"`
	CreatedAt         time.Time  `gorm:"not null;default:now()" json:"createdAt"`
	UpdatedAt         time.Time  `gorm:"not null;default:now()" json:"updatedAt"`

	User            User  `gorm:"foreignKey:UserID" json:"-"`
	CertificateFile *File `gorm:"foreignKey:CertificateFileID" json:"certificateFile,omitempty"`
}
